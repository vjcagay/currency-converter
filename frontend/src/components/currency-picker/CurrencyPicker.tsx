import React from "react";
import styles from "./styles.module.css";

const currencies = [
  {
    label: "ðŸ‡ºðŸ‡¸ USD - US Dollar",
    ticker: "USD",
  },
  {
    label: "ðŸ‡¸ðŸ‡¬ SGD - Singapore Dollar",
    ticker: "SGD",
  },
  {
    label: "ðŸ‡¯ðŸ‡µ JPY - Japanese Yen",
    ticker: "JPY",
  },
  {
    label: "ðŸ‡ªðŸ‡º EUR - Euro",
    ticker: "EUR",
  },
];

type Props = {
  from: string;
  to: string;
  onFromChange: (currency: string) => void;
  onToChange: (currency: string) => void;
};

const CurrencyPicker = (props: Props) => {
  return (
    <div className={styles.wrapper}>
      <label className={styles.label}>
        <span className={styles.labelText}>From</span>
        <select
          className={styles.select}
          data-testid="currency-picker-from"
          value={props.from}
          onChange={(event) => props.onFromChange(event.target.value)}
        >
          {currencies.map((currency) => (
            <option key={currency.ticker} value={currency.ticker}>
              {currency.label}
            </option>
          ))}
        </select>
      </label>
      <button
        data-testid="currency-picker-reverse"
        className={styles.button}
        onClick={() => {
          props.onFromChange(props.to);
          props.onToChange(props.from);
        }}
      />
      <label className={styles.label}>
        <span className={styles.labelText}>To</span>
        <select
          className={styles.select}
          data-testid="currency-picker-to"
          value={props.to}
          onChange={(event) => props.onToChange(event.target.value)}
        >
          {currencies.map((currency) => (
            <option key={currency.ticker} value={currency.ticker}>
              {currency.label}
            </option>
          ))}
        </select>
      </label>
    </div>
  );
};

export default CurrencyPicker;
