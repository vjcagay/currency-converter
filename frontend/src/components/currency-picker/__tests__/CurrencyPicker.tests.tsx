import { fireEvent, render } from "@testing-library/react";
import React from "react";
import CurrencyPicker from "../CurrencyPicker";

describe("<CurrencyPicker />", () => {
  it("should render correctly", () => {
    const { asFragment } = render(<CurrencyPicker from="USD" to="JPY" onChange={() => {}} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should invoke callbacks when selecting different values", async () => {
    const onChangeMock = jest.fn();

    const { findByTestId, rerender } = render(<CurrencyPicker from="USD" to="JPY" onChange={onChangeMock} />);
    fireEvent.change(await findByTestId("currency-picker-from"), { target: { value: "JPY" } });
    expect(onChangeMock).toHaveBeenCalledWith({ from: "JPY", to: "JPY" });

    rerender(<CurrencyPicker from="JPY" to="JPY" onChange={onChangeMock} />);
    fireEvent.change(await findByTestId("currency-picker-to"), { target: { value: "USD" } });
    expect(onChangeMock).toHaveBeenCalledWith({ from: "JPY", to: "USD" });
  });

  it("should reverse props values when clicking the reverse button", async () => {
    const onChangeMock = jest.fn();

    const { findByTestId } = render(<CurrencyPicker from="USD" to="JPY" onChange={onChangeMock} />);

    fireEvent.click(await findByTestId("currency-picker-reverse"));

    expect(onChangeMock).toHaveBeenCalledWith({ from: "JPY", to: "USD" });
  });
});
