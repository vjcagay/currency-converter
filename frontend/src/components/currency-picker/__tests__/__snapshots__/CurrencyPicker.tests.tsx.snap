// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CurrencyPicker /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <label
      class="label"
    >
      <span
        class="labelText"
      >
        From
      </span>
      <select
        class="select"
        data-testid="currency-picker-from"
      >
        <option
          value="USD"
        >
          ðŸ‡ºðŸ‡¸ USD - US Dollar
        </option>
        <option
          value="SGD"
        >
          ðŸ‡¸ðŸ‡¬ SGD - Singapore Dollar
        </option>
        <option
          value="JPY"
        >
          ðŸ‡¯ðŸ‡µ JPY - Japanese Yen
        </option>
        <option
          value="EUR"
        >
          ðŸ‡ªðŸ‡º EUR - Euro
        </option>
      </select>
    </label>
    <button
      class="button"
      data-testid="currency-picker-reverse"
    />
    <label
      class="label"
    >
      <span
        class="labelText"
      >
        To
      </span>
      <select
        class="select"
        data-testid="currency-picker-to"
      >
        <option
          value="USD"
        >
          ðŸ‡ºðŸ‡¸ USD - US Dollar
        </option>
        <option
          value="SGD"
        >
          ðŸ‡¸ðŸ‡¬ SGD - Singapore Dollar
        </option>
        <option
          value="JPY"
        >
          ðŸ‡¯ðŸ‡µ JPY - Japanese Yen
        </option>
        <option
          value="EUR"
        >
          ðŸ‡ªðŸ‡º EUR - Euro
        </option>
      </select>
    </label>
  </div>
</DocumentFragment>
`;
